### PR TITLE
Added `nats sub {{pattern}} --report-subjects`

### PR DESCRIPTION
**Pull request by davedotdev**
This handy feature (similar to `nats traffic`) builds a table of active subjects currently receiving data. Great for troubleshooting and exploration.

Better than running `nats sub ">"`!

### CODE REVIEW

File: cli/sub_command.go [hunk 1]
```diff
@@ -243,6 +308,44 @@ func (c *subCmd) subscribe(p *fisk.ParseContext) error {
 	}
 
 	switch {
+	case c.reportSubjects:
+		sub, err = nc.Subscribe(c.subject, subjectReportHandler)
+
+		go func() {
```
Comment (ripienaar) [inline cli/sub_command.go:314]: might be worth pulling this into a seperate function, this one is becoming (long ago) a monster

**Review by ripienaar — CHANGES_REQUESTED** [review]
Looks pretty good, few minor comments and also:

 * Remove the commented out code
 * Lets also track cumalative bytes per subject and report that
 * Lets add totals to the report

File: cli/sub_command.go [hunk 2]
```diff
@@ -209,6 +215,61 @@ func (c *subCmd) subscribe(p *fisk.ParseContext) error {
 		}
 	}
 
+	subjectReportHandler := func(m *nats.Msg) {
```
[the remark this answers is not in the document]
Comment (davedotdev) [reply cli/sub_command.go]: See: https://github.com/nats-io/natscli/pull/738/commits/aa3535a7678763bcb8998bade29a6a23dfe17a49

File: cli/sub_command.go [hunk 3]
```diff
@@ -243,6 +308,44 @@ func (c *subCmd) subscribe(p *fisk.ParseContext) error {
...
+					ticker.Stop()
+				default:
+					subjectRows := [][]any{}
+
+					if runtime.GOOS != "windows" {
+						fmt.Print("\033[2J")
+						fmt.Print("\033[H")
+					}
+
+					keys := make([]string, 0, len(subjectReportMap))
```
[the remark this answers is not in the document]
Comment (davedotdev) [reply cli/sub_command.go]: See: https://github.com/nats-io/natscli/pull/738/commits/aa3535a7678763bcb8998bade29a6a23dfe17a49

File: cli/sub_command.go (hunk 1 above)
Comment (davedotdev) [reply cli/sub_command.go]: See: https://github.com/nats-io/natscli/pull/738/commits/aa3535a7678763bcb8998bade29a6a23dfe17a49

File: cli/sub_command.go [hunk 4]
```diff
@@ -77,12 +81,67 @@ func configureSubCommand(app commandHost) {
...
 
 func init() {
 	registerCommand("sub", 17, configureSubCommand)
 }
 
+func startSubjectReporting(ctx context.Context, subjMu *sync.Mutex, subjectReportMap map[string]int64, subjectBytesReportMap map[string]int64, subjCount int) {
+	go func() {
+		ticker := time.NewTicker(time.Second)
+
+		for range ticker.C {
```
Comment (ripienaar) [inline cli/sub_command.go:96]: better to do like

```
for {
  select {
    case <-ticker.C:
    case ctx.Done():
  }
}
```

File: cli/sub_command.go [hunk 5]
```diff
@@ -77,12 +81,67 @@ func configureSubCommand(app commandHost) {
...
+					subjectRows = append(subjectRows, []any{k, subjectReportMap[k], subjectBytesReportMap[k]})
+					totalCount += subjectReportMap[k]
+					totalBytes += subjectBytesReportMap[k]
+					if (count + 1) == subjCount {
+						break
+					}
+				}
+				subjMu.Unlock()
+
+				table := newTableWriter("Active Subject Report")
```
Comment (ripienaar) [inline cli/sub_command.go:133]: if the count is > 10 in total then include the count somehow like `Top 10 Active Subjects Report`, this way its clear to people its a limited report and they know to go look for the flag to adjust the limit.

File: cli/sub_command.go [hunk 6]
```diff
@@ -77,12 +81,67 @@ func configureSubCommand(app commandHost) {
...
+				totalCount := int64(0)
+
+				subjMu.Lock()
+				keys := make([]string, 0, len(subjectReportMap))
+
+				for k := range subjectReportMap {
+					keys = append(keys, k)
+				}
+				// sort.Strings(keys)
+				// sort by count in descending order
```
Comment (ripienaar) [inline cli/sub_command.go:118]: not long before someone asks to select sorting by count or bytes :) for now ok lets wait for someone to ask

File: cli/sub_command.go [hunk 7]
```diff
@@ -77,12 +81,67 @@ func configureSubCommand(app commandHost) {
...
+				sort.Slice(keys, func(i, j int) bool {
+					return subjectReportMap[keys[i]] > subjectReportMap[keys[j]]
+				})
+
+				for count, k := range keys {
+					subjectRows = append(subjectRows, []any{k, subjectReportMap[k], subjectBytesReportMap[k]})
```
Comment (ripienaar) [inline cli/sub_command.go:124]: lets run subject count through humanize.Comma() and bytes through humanize.IBytes()

File: cli/sub_command.go [hunk 8]
```diff
@@ -77,12 +81,67 @@ func configureSubCommand(app commandHost) {
...
+					totalBytes += subjectBytesReportMap[k]
+					if (count + 1) == subjCount {
+						break
+					}
+				}
+				subjMu.Unlock()
+
+				table := newTableWriter("Active Subject Report")
+				table.AddHeaders("Subject", "Message Count", "Bytes")
+				table.AddFooter("Totals", totalCount, totalBytes)
```
Comment (ripienaar) [inline cli/sub_command.go:135]: humanize these numbers also

**Review by ripienaar — CHANGES_REQUESTED** [review]
looking really good, a few minor changes left :)

File: cli/sub_command.go (hunk 8 above)
Comment (davedotdev) [reply cli/sub_command.go:135]: See last comment! Hah.

File: cli/sub_command.go (hunk 4 above)
Comment (davedotdev) [reply cli/sub_command.go]: Done: https://github.com/nats-io/natscli/pull/738/commits/e49fd63a0929654078661894fad1d5313a8be3a4

File: cli/sub_command.go (hunk 5 above)
Comment (davedotdev) [reply cli/sub_command.go]: Done: https://github.com/nats-io/natscli/pull/738/commits/e49fd63a0929654078661894fad1d5313a8be3a4

**Comment by davedotdev** [comment]
Addressed final run of comments/asks/issues:
```
╭────────────────────────────────────╮
│    Top 10 Active Subjects Report   │
├──────────┬───────────────┬─────────┤
│ Subject  │ Message Count │ Bytes   │
├──────────┼───────────────┼─────────┤
│ 12vprobe │ 698           │ 4.8 KiB │
│ 24vprobe │ 282           │ 1.9 KiB │
│ barrier  │ 50            │ 500 B   │
├──────────┼───────────────┼─────────┤
│ TOTALS   │ 1,030         │ 7.2 KIB │
╰──────────┴───────────────┴─────────╯
```

**Pull request merged**